### ISSUE TITLE
NetBox mixes authors of modifications in the changelog when there are a lot of concurrent modifications

### ISSUE THREAD

**Issue by drygdryg** [netbox-community/netbox#7696]
### NetBox version

v3.0.8

### Python version

3.9

### Steps to Reproduce

1. Set up a test local instance of NetBox.
2. Download the necessary files: https://gist.github.com/drygdryg/f32038ef94514aa5503098dbbb42a46c
3. Restore backup of the database of my NetBox test instance (`netbox.sql`) to your NetBox instance [according the manual](https://netbox.readthedocs.io/en/stable/administration/replicating-netbox/).
4. Edit `test_case.py`: configure `NETBOX_URL` according to your NetBox instance.
5. Run `test_case.py`. You need to install the httpx module to run if it's not installed: `pip install httpx`.
6. Log in to NetBox using the following credentials:
username: admin
password: admin
7. View Changelog (`http://netbox.local/extras/changelog/`).

### Expected Behavior

According to the changelog, each user should change his device based on this logic:
`test_user1` updates `Test router 1`
`test_user2` updates `Test router 2`
`test_user3` updates `Test router 3`
etc.
### Observed Behavior

Users are mixed in the changelog. Some examples:
![image](https://user-images.githubusercontent.com/43933400/139524544-13b47060-98f6-41fc-8ae5-573ab74058e1.png)
![image](https://user-images.githubusercontent.com/43933400/139524563-a83c1d0d-a30f-47b1-a7a8-20be8a5f0676.png)

**Comment by kkthxbye-code** [comment]
Probably same issue as #7657, #5142, #5609 - There's a race condition in ObjectChange handling when using an application server running multiple workers in threads.

I don't really know a lot about django, but isn't this iffy with the yield? I'm not sure that is thread safe.

https://github.com/netbox-community/netbox/blob/0b0ab9277cdb8f388f2d6aee92d708d5126b9111/netbox/extras/context_managers.py#L25-L38

**Comment by tyler-8** [comment]
@drygdryg Can you try this branch and see if it resolves the issue for you? It seems to fix it on my dev instance. https://github.com/tyler-8/netbox/tree/objectchange_threadsafe

**Comment by tyler-8** [comment]
I made a new update, try the https://github.com/tyler-8/netbox/tree/objectchange_threadsafe branch now.

**Comment by drygdryg** [comment]
> @drygdryg Can you try this branch and see if it resolves the issue for you? It seems to fix it on my dev instance. https://github.com/tyler-8/netbox/tree/objectchange_threadsafe

@tyler-8 I tested this with 9 workers and 3 threads, and it solved the issue for me, thanks.

**Comment by tyler-8** [comment]
> @tyler-8 I tested this with 9 workers and 3 threads, and it solved the issue for me, thanks.

Try fewer workers and more threads; such as 4 and 4.

For testing this, we want the threads to be used more heavily and with 9 workers (processes) against your test script there is barely enough load for threads.

**Comment by DanSheps** [comment]
Closing this in favour of #7657

**Comment by tyler-8** [comment]
I'm experimenting a slightly different track with this one now @DanSheps - this particular issue might be more due to using non-unique values for `dispatch_uid` in the signal `.connect()` methods when using `gunicorn` threading. Related to the other issue, but possibly different enough to warrant another solution.

https://github.com/netbox-community/netbox/blob/b5e81577001f04f489e9fb000d4229e7b7cc9733/netbox/extras/context_managers.py#L26

would become

```python
post_save.connect(handle_changed_object, dispatch_uid=f'handle_changed_object_{request.id}')
```

**Comment by tyler-8** [comment]
@drygdryg - can you try again using this branch instead? https://github.com/tyler-8/netbox/tree/objectchange_dispatch_uids

I'm seeing good results with your script without the use of `threading.Lock()` like in the previous branch

**Comment by DanSheps** [comment]
@tyler-8 since this one might be semi-unrelated, I will re-open this for you

**Comment by tyler-8** [comment]
Boy this is frustrating to troubleshoot and test, lol. 

Initially, using truly unique `dispatch_uid` seemed to work through a couple test runs of the `test_case.py` script, but now I'm seeing the user-mixups. On the other hand the original branch https://github.com/tyler-8/netbox/tree/objectchange_threadsafe is still solving the issue consistently.

**Comment by tyler-8** [comment]
Benchmarks using a modified `test_case.py` (ThreadPoolExecutor instead of ProcessPoolExecutor)

```
Client 7 threads, 1 user per thread
gunicorn 8 processes (essentially, no thread-locks)
1:09.33
1:03.45
1:00.36
1:00.26
0:55.067
1:00.82

Client 7 threads, 1 user per thread
gunicorn 2 processes, 5 threads
1:40.41
1:40.45
1:43.61
1:41.18
1:39.32
1:40.59

Client 7 threads, 1 user per thread
gunicorn 5 processes, 3 threads
1:19.74
1:04.45
1:05.68
1:10.65
1:08.53
1:04.65

Client 7 threads, 1 user per thread
gunicorn 4 processes, 4 threads
1:02.83
1:02.15
1:44.20
0:59.19
1:22.28
1:00.38
```

**Comment by tyler-8** [comment]
@DanSheps After tinkering with this issue more today (and digging up the old issues that had similar errors/behavior), I think it is in fact the same issue. You can close this one again (sorry :P ) and I'll post my updates in #7657

**Comment by jeremystretch** [comment]
No worries! We'll close this one for now, and revisit if the eventual fix for #7657 is determined not to resolve this particular issue as well.
